### PR TITLE
Double click on entity in blueprint & timepanel focuses camera on it in 3D space views

### LINKED ISSUE #917: viewer: double clicking an entity in the blueprint should put it into focus
(Following a standup discussion)

For spatial entities, double-clicking on their name in the blueprint should:
1. make their centroid the focus of the orbit camera
1. zoom out the camera so that the entire entity can fit on screen

Both can be achieved relatively easily by computing bounding spheres (e.g. [here's how Bevy's scene_viewer handles this](https://github.com/bevyengine/bevy/blob/v0.9.1/examples/tools/scene_viewer.rs#L214-L296)), though this might lead to potentially undesirable edge cases in some very specific scenarios (very oddly shaped meshes, point clouds with extreme outliers, etc).

## Future work

Possible future work mentioned during the discussion:
- Could offer a way for end-users to specify custom centroids/focus-points and optimal zoom-out distances using e.g. dedicated components

### LINKED ISSUE #918: viewer: double clicking an entity in the spatial view should put it into focus
(Following a standup discussion)

The controversial follow-up to #917: generally the same idea, except that it would also work when directly double-clicking on an entity through the spatial view.

This requires some design work; some ideas/thoughts that were thrown around:
- only do so if the entity was already selected to begin with
- what's the expected with e.g. point clouds (i.e. how does one click on a point cloud vs. a point in the point cloud)

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #918
* Fixes #917

This PR introduces on-the-fly determined per-entity bounding boxes and unifies a lot of behavior under the hood: A double click of a hovered item (pretty much everywhere) causes now a "focused" event for that item. Focuses are right now set on the viewer context for a single frame, so each part of the viewer can react on it as it pleases. The 3D Space view reacts by focusing the camera on the entity (using the new bounding box) and tracks cameras.
Tracking of entities has been unified - originally this PR tracked entities as well, but I found this a bit too confusing for this action, we should instead have something more explicit (this was disabled in 86ddf0d51e29803fa93da1608fa884f0fc0496ca so it's easy to get it back).

https://github.com/rerun-io/rerun/assets/1220815/b6077865-d500-4bc9-b112-8ff41b370fe0

Tradeoffs:
* can't focus on instances
    -> before focused on 3D positions (without knowing anything about the object under it) now focusing on entities; different behavior, _sometimes_ worse, sometimes better
    -> note that tracking an instance rarely what you want here (e.g. point instance ids vary widely)
* tracking an entity is not always what you want -> disabled it
    * it's very hard to tell that we're tracking something, can sometimes be surprising
* Alternative design of `focus`: Send a focus event to space views that should be interested in it. This makes would unify the dispatching code for this which is nicer going forward. But for the moment it was just a lot easier to implement it the way I did. I'm fairly sure though about making focus not sticky in the global state since this way we'd end up with two levels of selection.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4799/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4799/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4799/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4799)
- [Docs preview](https://rerun.io/preview/7183423673682ce3d9c0c68fcb335689b8c13693/docs) 
- [Examples preview](https://rerun.io/preview/7183423673682ce3d9c0c68fcb335689b8c13693/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_space_view_spatial/src/ui_3d.rs [hunk 1]
```diff
@@ -444,53 +508,60 @@ pub fn view_3d(
...
+                if space_view.is_none() || space_view == &Some(query.space_view_id) {
+                    Some(&path.entity_path)
+                } else {
+                    None
+                }
+            }
+        };
+        if let Some(entity_path) = focused_entity {
+            // For the moment we only track cameras.
+            // We have everything in place to track arbitrary objects but it's a bit odd sometimes, so only focus on them instead.
```
Comment (emilk) [inline crates/re_space_view_spatial/src/ui_3d.rs:539]: By "focus" you mean, move camera to them now, but don't track over time?

Comment (emilk) [inline crates/re_space_view_spatial/src/ui_3d.rs:539]: Odd how? What need to change for us to want to start tracking all entities?

File: crates/re_space_view_spatial/src/ui_3d.rs [hunk 2]
```diff
@@ -444,53 +508,60 @@ pub fn view_3d(
...
+            if space_cameras.iter().any(|c| &c.ent_path == entity_path) {
```
Comment (emilk) [inline crates/re_space_view_spatial/src/ui_3d.rs:540]:
```suggestion
            let is_camera = space_cameras.iter().any(|c| &c.ent_path == entity_path);
            if is_camera {
```

File: crates/re_viewer_context/src/selection_state.rs [hunk 3]
```diff
@@ -27,7 +27,7 @@ pub enum SelectedSpaceContext {
 
         /// Path of a space camera, this 3D space is viewed through.
         /// (None for a free floating Eye)
```
Comment (emilk) [inline crates/re_viewer_context/src/selection_state.rs:29]: outdated docstring

**Review by emilk — APPROVED** [review]

File: crates/re_space_view_spatial/src/scene_bounding_boxes.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        for visualizer in visualizers.iter() {
+            if let Some(data) = visualizer
+                .data()
+                .and_then(|d| d.downcast_ref::<SpatialViewVisualizerData>())
+            {
+                for (entity, bbox) in &data.bounding_boxes {
+                    self.per_entity
+                        .entry(*entity)
+                        .and_modify(|bbox_entry| *bbox_entry = bbox_entry.union(*bbox))
+                        .or_insert(*bbox);
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_space_view_spatial/src/scene_bounding_boxes.rs:45]: I haven't measured before/after but I don't have reason to believe that this is going to make much of a difference: Previously we accumulated them all into the scene bounding box directly, now we go through an indirection and collect them in this method

At some point we'll have to add a test scene with thousands of entitites to get a sense for these kind of changes.

File: crates/re_space_view_spatial/src/ui_3d.rs (hunk 1 above)
Comment (Wumpf) [reply crates/re_space_view_spatial/src/ui_3d.rs:539]: yes, will clarify the comment

**Comment by Wumpf** [comment]
when testing one last time on the web I found that camera tracking lags behind, so this regressed

File: crates/re_space_view_spatial/src/scene_bounding_boxes.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_space_view_spatial/src/scene_bounding_boxes.rs:45]: We used to do `aabb += aabb` for each entity, but now we also store and lookup each such aabb in a hash table. That sounds a significantly more expensive to me.

Yes, we should have a test for this - maybe whip up a quick one, e.g. just generate 10k individual points?

**Comment by Wumpf** [comment]
added a primitive benchmark that adds lots of entitites. No measurable regression, but it's shocking how few entities we can handle

**Pull request merged**